syntax = "proto3";

import "google/protobuf/empty.proto";

package chat;

service ChatService {
  rpc Connect(ConnectRequest) returns (ConnectResponse);
  rpc Disconnect(DisconnectRequest) returns (google.protobuf.Empty);
  rpc SendMessage(SendMessageRequest) returns (google.protobuf.Empty);
  rpc SubscribeMessages(InformClientsNewMessageRequest)
      returns (stream InformClientsNewMessageResponse);
  rpc SubscribeClientEvents(google.protobuf.Empty)
      returns (stream ClientEventData);
}

message ClientEventData{

  enum ClientEventType {
  ADD = 0;
  REMOVE = 1;
  SYNC = 2;
  }

  repeated string pseudonyms = 1;
  ClientEventType event_type = 2;
}

message ConnectRequest {
  reserved 1;
  string pseudonym = 2;
  string gender = 3;
  string country = 4;
}

message ConnectResponse {
  bool accepted = 1;
  string message = 2;
  repeated string previous_message_context = 3;
}

message DisconnectRequest {
  string pseudonym = 1;
}

message SendMessageRequest {
  string content = 1;
  reserved 2;
}

message InformClientsNewMessageRequest {
  reserved 1;
}

message InformClientsNewMessageResponse {
  string author = 1;
  string content = 2;
}
